`best_blame_constraint`: Blame better constraints when the region graph has cycles from invariance or `'static`

dianne opened the pull request:
This fixes #132749 by changing which constraint is blamed for region errors in several cases. `best_blame_constraint` had a heuristic that tried to pinpoint the constraint causing an error by filtering out any constraints where the outliving region is unified with the ultimate target region being outlived. However, it used the SCCs of the region graph to do this, which is unreliable; in particular, if the target region is `'static`, or if there are cycles from the presence of invariant types, it was skipping over the constraints it should be blaming. As is the case in that issue, this could lead to confusing diagnostics. The simplest fix seems to work decently, judging by test stderr: this makes `best_blame_constraint` no longer filter constraints by their outliving region's SCC.

There are admittedly some quirks in the test output. In many cases, subdiagnostics that depend on the particular constraint being blamed have either started or stopped being emitted. After starting at this for quite a while, I think anything too fickle about whether it outputs based on the particular constraint being blamed should instead be looking at the constraint path as a whole, similar to what's done for [the placeholder-from-predicate note](https://github.com/rust-lang/rust/compare/master...dianne:rust:better-blame-constraints-for-static#diff-3c0de6462469af483c9ecdf2c4b00cb26192218ef2d5c62a0fde75107a74caaeR506).

Very many tests involving invariant types gained a note pointing out the types' invariance, but in a few cases it was lost. A particularly illustrative example is [tests/ui/lifetimes/copy_modulo_regions.stderr](https://github.com/rust-lang/rust/compare/master...dianne:rust:better-blame-constraints-for-static?expand=1#diff-96e1f8b29789b3c4ce2f77a5e0fba248829b97ef9d1ce39e7d2b4aa57b2cf4f0); I'd argue the new constraint is a better one to blame, but it lacks the variance diagnostic information that's elsewhere in the constraint path. If desired, I can try making that note check the whole path rather than just the blamed constraint.

The subdiagnostic [`BorrowExplanation::add_object_lifetime_default_note`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_borrowck/diagnostics/explain_borrow/enum.BorrowExplanation.html#method.add_object_lifetime_default_note) depends on a `Cast` being blamed, so [a special case](https://github.com/rust-lang/rust/pull/133858/commits/364ca7f99c12fb5220e6b568ac391979317ce878) was necessary to keep it from disappearing from tests specifically testing for it. However, see the FIXME comment in that commit; I think the special case should be removed once that subdiagnostic works properly, but it's nontrivial enough to warrant a separate PR. Incidentally, this removes the note from a test where it was being added erroneously: in [tests/ui/borrowck/two-phase-surprise-no-conflict.stderr](https://github.com/rust-lang/rust/compare/master...dianne:rust:better-blame-constraints-for-static?expand=1#diff-8cf085af8203677de6575a45458c9e6b03412a927df879412adec7e4f7ff5e14), the object lifetime is explicitly provided and it's not `'static`.

dianne committed:
`best_blame_constraint`: don't filter constraints by sup SCC

The SCCs of the region graph are not a reliable heuristic to use for blaming an interesting
constraint for diagnostics. For region errors, if the outlived region is `'static`, or the involved
types are invariant in their lifetiems, there will be cycles in the constraint graph containing both
the target region and the most interesting constraints to blame. To get better diagnostics in these
cases, this commit removes that heuristic.

nnethercote commented:
I am not a good person to review nitty gritty borrowck error message changes :)  Let's try someone else who is probably more appropriate: r? @lcnr

> this may have perf consequence

Let's check: @bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 189b2f892e6d0809a77fc92fe1108a07d8de9be0 with merge 889e047c4bb3d82cd34c0c703f2bd13964c828ca...
<!-- homu: {"type":"TryBuildStarted","head_sha":"189b2f892e6d0809a77fc92fe1108a07d8de9be0","merge_sha":"889e047c4bb3d82cd34c0c703f2bd13964c828ca"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12169129217/job/33945293256)
Build commit: 889e047c4bb3d82cd34c0c703f2bd13964c828ca (`889e047c4bb3d82cd34c0c703f2bd13964c828ca`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12169129217/job/33945293256"},"merge_sha":"889e047c4bb3d82cd34c0c703f2bd13964c828ca"} -->

rust-timer commented:
Queued 889e047c4bb3d82cd34c0c703f2bd13964c828ca with parent acabb5248231987ae1f0c215208d1005a5db402d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=acabb5248231987ae1f0c215208d1005a5db402d&end=889e047c4bb3d82cd34c0c703f2bd13964c828ca).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (889e047c4bb3d82cd34c0c703f2bd13964c828ca): [comparison URL](https://perf.rust-lang.org/compare.html?start=acabb5248231987ae1f0c215208d1005a5db402d&end=889e047c4bb3d82cd34c0c703f2bd13964c828ca&stat=instructions:u).

### Overall result: no relevant changes - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=acabb5248231987ae1f0c215208d1005a5db402d&end=889e047c4bb3d82cd34c0c703f2bd13964c828ca&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=acabb5248231987ae1f0c215208d1005a5db402d&end=889e047c4bb3d82cd34c0c703f2bd13964c828ca&stat=max-rss)
<details>
<summary>Results (secondary 2.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4% | [2.4%, 2.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=acabb5248231987ae1f0c215208d1005a5db402d&end=889e047c4bb3d82cd34c0c703f2bd13964c828ca&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=acabb5248231987ae1f0c215208d1005a5db402d&end=889e047c4bb3d82cd34c0c703f2bd13964c828ca&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 769.678s -> 767.43s (-0.29%)
**Artifact size**: 330.86 MiB -> 330.84 MiB (-0.01%)

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -2168,17 +2175,38 @@ impl<'tcx> RegionInferenceContext<'tcx> {
+                        }
+                )
+            {
+                // FIXME(dianne): `BorrowExplanation::add_object_lifetime_default_note` depends on a
+                // coercion being blamed, so revert to the old blaming logic to prioritize that.
+                // The note's logic should be reworked, though; it's flaky (#131008 doesn't have a
+                // coercion, and even with this hack, one isn't always blamed when present).
+                // Only checking for a coercion also makes the note appear where it shouldn't
+                // shouldn't (e.g. `tests/ui/borrowck/two-phase-surprise-no-conflict.stderr`).
```
from what I can tell, this FIXME is stating that this branch is causing us to not emit a `add_object_lifetime_default_note` in cases where we should. Please link to an affected test

dianne replied:
ah, the wording of the FIXME is unclear; I'll see if I can make it better. I've made sure to still emit that note where it was previously emitted. that branch is a hack specifically to keep the note present in [tests/ui/traits/trait-object-lifetime-default-note.stderr](https://github.com/dianne/rust/blob/better-blame-constraints-for-static/tests/ui/traits/trait-object-lifetime-default-note.stderr), since otherwise the `CallArgument` constraint would be blamed and the note would disappear; it'll be more apparent once I get rid of the `UseAsConst`/`UseAsStatic` special-casing and split the `Cast` case into a separate commit. I'll refer to that test in the FIXME too.

the missing notes the FIXME is referring to are cases where the note would make sense (and imo should be emitted)[^1], but it wasn't emitted even before this change. the note depends on a coercion being blamed, but it won't always be blamed for region errors from object lifetime defaults (nor will a coercion necessarily be present). handling that properly is (afaict) nontrivial; I'd be happy to do what I can there in a separate PR, but I might either need some help with wording or some direction on what heuristics would make sense, since doing it as perfectly as possible seems quite involved.

[^1]: see #131008, which would probably need a fairly general solution. I remember finding in-tree tests that are missing the note too though (which would require a less-general but still nontrivial fix), so I can dig those up if it'd help

lcnr commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -2168,17 +2175,38 @@ impl<'tcx> RegionInferenceContext<'tcx> {
+            // If an "outlives 'static" constraint was from use as a const or static, blame that.
+            if target_is_static
+                && blame_source
+                && let Some(old_best) = categorized_path.iter().min_by_key(|p| p.category)
```
this feels odd to me, do you want to disable this if there's also one step in the path whose category is a `ConstraintCategory::Return` or would it be pattern to walk the path to detect whether any `ConstraintCategory::UseAsConst` exists?

dianne replied:
looking at it again, that special case shouldn't be there at all; it's just papering over a more fundamental issue. from what I can tell, the reason [tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr](https://github.com/dianne/rust/blob/189b2f892e6d0809a77fc92fe1108a07d8de9be0/tests/ui/consts/const-mut-refs/mut_ref_in_final.stderr#L38) needs the special-casing is that `UnsafeCell<T>` is invariant in `T`, hence the outliving region of the `UseAsConst`/`UseAsStatic` constraint is in the same SCC as the target, hence it gets skipped. for a quick workaround, this could check if there's a constraint on the path with an invariant `variance_info` before removing any edges from the region graph. it's sort of justifiable, since back-edges from invariance would also need to be removed in order to get the properties I wanted from pruning the `'static: R` edges. that said, I think this should handle invariance properly if it can. leaving it sometimes-wrong feels pretty unsatisfying, especially if a proper fix might handle this whole issue cleanly. I'll have to look into it further.

lcnr reviewed (commented):
I think this mostly makes sense to me, it has been somewhat difficult to review for me however. It would be really nice if you could clean up the surrounding code for a bite, even though it's not trictly part of your change. This would make reviewing this code a lot more pleasant.


- afaict `extra_info` is only mutated by a single `push` in this function, so 1) why are we using a `Vec` instead of an `Option` here and 2) could we its computation into a separate function
- I can't immediately tell what changes due to removing the `'whatever: 'static` edges and what changes due to the explicit check+eager return, can you give me some more info here, or maybe even split it into 2 commits?

dianne commented:
> * afaict `extra_info` is only mutated by a single `push` in this function, so 1) why are we using a `Vec` instead of an `Option` here and 2) could we its computation into a separate function

I think the intent is for there to possibly be more variants of `ExtraConstraintInfo`. I agree though; think it'd be cleaner to move it to a separate function, get rid of the `Vec`, and make it a `struct`. I'll try that!

> * I can't immediately tell what changes due to removing the `'whatever: 'static` edges and what changes due to the explicit check+eager return, can you give me some more info here, or maybe even split it into 2 commits?

no problem! I'll split it up; that'll help me too. I'll also be removing the special-case for `UseAsConst`/`UseAsStatic` as per [my reply to your review comment](https://github.com/rust-lang/rust/pull/133858#discussion_r1879483388), so hopefully that'll remove some noise.

dianne committed:
cleanup: remove `ExtraConstraintInfo`

`ExtraConstraintInfo` was used only for a single subdiagnostic, so this moves the logic for that
to its own function and eliminates the indirection. In order to do so cleanly, this also changes
the arguments to `BorrowExplanation::add_explanation_to_diagnostic`, which happens to simplify its
call sites.

dianne committed:
further clean up `best_blame_constraint`

This gets rid of `categorized_path`, as it was redundant given the `OutlivesConstraint`s in `path`
already have a category field.

dianne committed:
`best_blame_constraint`: add a special case to recover object lifetime default notes

dianne commented:
I've pushed some changes:
- The first two commits are cleanup of the surrounding code.
- The third is the primary diagnostic change, which I've reworked. I'll try and write up a revised description in the main PR text, but in short, I've removed all of the logic to filter constraints by their sup's SCC. The result is simpler and I think it produces better diagnostics (though a few bits could use some work; I'll give my opinion in the longer writeup).
- The fourth commit is the special case for `tests/ui/traits/trait-object-lifetime-default-note.rs`, including an updated FIXME comment.

bors commented:
:umbrella: The latest upstream changes (presumably #132706) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr reviewed (commented):
r=me on the first two commits, some of the changes look like they make things worse :thinking: have to look into it in more detail

thanks for splitting this PR into these commits, I feel like it's a lot more pleasant to review :heart:

lcnr commented on tests/ui/async-await/issue-76547.stderr:
```diff
@@ -1,27 +1,25 @@
+   |              ^^^^  -     - let's call the lifetime of this reference `'2`
+   |              |     |
+   |              |     let's call the lifetime of this reference `'1`
+   |              assignment requires that `'1` must outlive `'2`
```
this seems wrong to me :thinking:

the assignment is very clearly not part of the function signature xx that's likely caused by the async fn desugaring first moving `bufs` into a local, and we then constrain the type of the local to be `&'a mut &'a` :/

I think we should somehow filter this, I feel like the current approach of figuring out what to blame is generally a bit :shrug: i would expect us to instead iterate over all constraints and take the "best one" (while preferring earlier ones), instead of simply taking the first interesting one

dianne replied:
ah, yeah, that's definitely wrong. I had a note about that test but I completely forgot to address it; thank you for the catch! I've added a commit that makes it try not to blame constraints from desugarings. it includes a special case to allow blaming returns from `?`, since otherwise [`tests/ui/async-await/issue-67765-async-diagnostic.stderr`](https://github.com/rust-lang/rust/blob/master/tests/ui/async-await/issue-67765-async-diagnostic.stderr) would be less clear about why the usage `Err(b)` is problematic (and `?` including a return is a lot more clear than the async sugar including an assignment). I'm not a desugaring expert, though, so I'm not sure if any others are intuitive/meaningful enough to be blamed (possibly with special wording?).

with regard to iterating over all constraints and picking the "best one", I agree that it should look at all the constraints in some way, but I'll have to think about if there's a better heuristic for the single best than position on the path (with some filter). I think returns are generally good to point to, and `CopyBound` constraints should maybe at least get an honorable mention in a note (if not the full blame) since they're subtle, but beyond that I think it gets a bit fuzzy. e.g. ascriptions and casts are helpful to point to when they syntactically refer to a relevant type/region, but are much less helpful if they leave it up to inference. for another example, `UseAsConst` is sometimes the best we've got at the top level, but within function bodies it's usually less helpful. besides the category, it feels like it should be possible to use constraints' regions to pinpoint the most blameable constraint, but I haven't found a straightforward way to get blame out of the constraint graph or region graph, given the cycles. constraints' location/span are helpful, but mostly once we're in proper diagnostic code and can look at types and the HIR and MIR and such. the compromise I've imagined is to pick a good-enough-but-maybe-not-best constraint to blame, and anything important enough to surface in addition to that can be picked out from the constraint path by diagnostics (though of course if there's some other way to get the blamed constraint as good as possible too, that'd be ideal).

dianne committed:
`best_blame_constraint`: avoid blaming constraints from MIR generated by desugaring

bors commented:
:umbrella: The latest upstream changes (presumably #134292) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lcnr reviewed (commented):
a final suggestion, apart from this r=me :thinking: 

borrowck diagnostics are :/

bors commented:
:umbrella: The latest upstream changes (presumably #134243) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne commented on tests/ui/match/match-ref-mut-invariance.stderr:
```diff
@@ -1,12 +1,12 @@
    |            -- lifetime `'a` defined here
 LL |         match self.0 { ref mut x => x }
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'b` but it is returning data with lifetime `'a`
+   |                        ^^^^^^^^^ assignment requires that `'a` must outlive `'b`
```
oh, I missed that! I believe those bindings are "assignment"s because once they're lowered to MIR locals, they're assigned to by MIR assignment statements; the constraint generated by that is what's getting blamed. afaict currently there's not enough information in the MIR to efficiently make that distinction when assigning constraint categories. but we do have a span, so it shouldn't be too hard to call assignments "binding"s in diagnostics when syntactically they feel more like a "binding" than an "assignment".

that said, those are the wrong constraints to blame for these tests; these are both regressions, though it's just by chance they were correct in the first place. I assume because the error depends on `&mut` being invariant in its parameter, we're looking at things kind of backwards[^1] and we picked a constraint from the unhelpful end of the path. I think it should be possible to detect when constraints arising from contravariance are being followed and flip the path around if so.. it'll probably make for better diagnostics, but I'll admit I don't love relying so heavily on what direction we traverse the constraint graph in for this. I'm worried something more robust might require collecting more information in borrowck, but I'll look into it at least.

sorry for the huge number of very specific stderr changes, and thank you for taking the time to review this. I definitely agree that it's :/

[^1]: this is present in this test when blaming the correct constraint too: the label "method was supposed to return data with lifetime `'b` but it is returning data with lifetime `'a`" is peculiar when the return type in the signature has lifetime `'a` and the body has lifetime `'b`. I think that message was worded assuming return types are always covariant. it should be fixed too (maybe with a few other tweaks as well.. ideally it'd suggest replacing `'a` with `'b` rather than adding `'a: 'b`, like we do for `'static`), but probably that should go in a different PR.

lcnr replied:
hmm, one approach would be to change this code here to consider more assignments to be boring: https://github.com/rust-lang/rust/blob/3378a5e084fa51c2e7f5c3f53b87e8ad89f1c517/compiler/rustc_borrowck/src/type_check/mod.rs#L1117-L1139 

E.g. by checking for `RefForGuard` and looking at `VarBindingForm::opt_match_place`. It's a bit fiddly sadly

lcnr replied:
One likely quite impactful suggestion: make it boring if `opt_ty_info` is `None`

dianne replied:
that's where I was looking, yeah. though I was focusing too much on distinguishing between assignment syntaxes. `opt_match_place` does work for catching those by-ref bindings. and since we (probably?) don't need the span or place from that, checking the `VarBindingForm`'s `binding_mode` should be equivalent (I think). I don't think they should all be boring though, so there'd at need to be some other check (if not a different check entirely). e.g., for [field-projection-mutating-context.rs](https://github.com/rust-lang/rust/blob/master/tests/ui/mir/field-projection-mutating-context.rs):
```diff
1	error[E0308]: mismatched types
-	   --> $DIR/field-projection-mutating-context.rs:9:13
+	   --> $DIR/field-projection-mutating-context.rs:9:25
3	    |
4	 LL |     let Foo(ref mut y): Foo<fn(&'static str)> = x;
-	    |             ^^^^^^^^^ one type is more general than the other
+	    |                         ^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
6	    |
7	    = note: expected fn pointer `for<'a> fn(&'a _)`
8	               found fn pointer `fn(&_)`
```
the assignment is what relates the two types, so it makes more sense to blame than just one of the type annotations. there's also [fn-subtype.stderr](https://github.com/rust-lang/rust/blob/master/tests/ui/nll/relate_tys/fn-subtype.stderr) specifically checking for an assignment being blamed, though that assignment is by-value. in both cases the span associated with the assignment's constraint is unfortunate, but it is the assignment being blamed.

`RefForGuard` locals being boring to assign to makes sense, but it doesn't look like we blame those assignments in any tests anyway. I'd have to think more to come up with an example, if it's possible.

and `opt_ty_info` isn't *quite* what the docs claim. it looks like it's only `Some(_)` for function inputs (specifically for the inputs themselves, not projections from destructuring patterns). any other binding gets an `opt_ty_info` of `None` even if there's an annotation on it. we can get user annotations on normal local variables from a local decl's `user_ty` field, though. there's a slight asymmetry in how `UserTypeProjections` accounts nicely for projections and `opt_ty_info` doesn't, but also I can't imagine ever wanting to point to constraints from an argument being destructured, so that's probably fine (if unsettlingly ad-hoc). just checking for those also misses assignments from paths with explicitly-provided generic arguments, but maybe that's fine too; I think getting all user types and sorting out what's inferred from what's provided would be a much too involved process. in most of the tests this affects, it doesn't matter too much whether the assignment or something else gets blamed; they each provide half of the reason for a region error, so unless we surface more constraints' spans, the user has to hunt down the other half themself either way.

dianne committed:
`best_blame_constraint`: avoid blaming assignments without user-provided types

dianne commented:
I've rebased and added an experiment in treating some assignments without sufficiently nearby type annotations as uninteresting. this reverts several tests' output to what they were prior to this PR.

lcnr commented:
I think the impact of b041c50d2dc2650e90b7f3b1bbe4bb955b5044b6 is positive :+1: you need to bless mir-opt tests

wrt to https://github.com/rust-lang/rust/pull/133858#discussion_r1884375062

I feel like both "the first constraint" and "more interesting constraint kinds" are important metrics when considering what to blame. In a sense maybe a good compromise would be:

Have different groups of "constraint priorities" and then choose the earliest constraint with the highest priority

This pretty much matches the current behavior except that there only exist 2 priority classes: boring and interesting :thinking: i won't require u to impl it in this PR, I do think something like this would allow for a nice-ish (as nice as it can get) solution here

dianne commented:
mir-opt tests should now be blessed. 

and I have something like that implemented locally already actually! I was experimenting with that as a way of making the fallback smarter for when an interesting constraint wasn't found due to cycles (back when that was confusing it). I'll see about updating that and checking how it runs tomorrow.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12410557563/job/34646454144) [(plain)](https://github.com/rust-lang/rust/commit/c895dce3509f761e91252a5d36615c23b1bd7d67/checks/34646454144/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.3s done
#21 DONE 32.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.95s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
................................................

failures:

---- [mir-opt] tests/mir-opt/nll/region_subtyping_basic.rs stdout ----
17 | '?2 live at {bb1[0]}
18 | '?3 live at {bb1[1..=3]}
19 | '?4 live at {bb1[4..=7], bb2[0..=2]}
- | '?2: '?3 due to Assignment at Single(bb1[0]) ($DIR/region_subtyping_basic.rs:19:13: 19:18 (#0)
- | '?3: '?4 due to Assignment at Single(bb1[3]) ($DIR/region_subtyping_basic.rs:20:13: 20:14 (#0)
+ | '?2: '?3 due to Assignment { has_interesting_ty: false } at Single(bb1[0]) ($DIR/region_subtyping_basic.rs:19:13: 19:18 (#0)
+ | '?3: '?4 due to Assignment { has_interesting_ty: false } at Single(bb1[3]) ($DIR/region_subtyping_basic.rs:20:13: 20:14 (#0)
23 | Borrows
23 | Borrows
24 | bw0: issued at bb1[0] in '?2
thread '[mir-opt] tests/mir-opt/nll/region_subtyping_basic.rs' panicked at src/tools/compiletest/src/runtest/mir_opt.rs:93:21:
thread '[mir-opt] tests/mir-opt/nll/region_subtyping_basic.rs' panicked at src/tools/compiletest/src/runtest/mir_opt.rs:93:21:
Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/nll/region_subtyping_basic.main.nll.0.32bit.mir


failures:
    [mir-opt] tests/mir-opt/nll/region_subtyping_basic.rs
```

</details>

dianne committed:
remove the unused `ConstraintCategory::ClosureBounds`

dianne committed:
`best_blame_constraint`: prioritize blaming interesting-seeming constraints

dianne committed:
make outlives constraints from generic arguments less boring

dianne committed:
make outlives constraints from pointer comparisons less boring

dianne committed:
point out unblamed constraints from `Copy`/`Sized` bounds in region errors

dianne committed:
only avoid blaming assignments from argument patterns

dianne commented:
I've tried ordering constraints by how interesting they are. after some experimentation, it looks like most categories should have the same priority, but there are exceptions. using the ordering of constraints on the path can still be unhelpful due to contravariance/invariance, but prioritizing blaming returns helps quite a bit (at least in several ui tests). a proper variance-respecting heuristic looks like it'd require some pretty substantial changes, so hopefully it's good enough just working around that.

a few more changes:
- pointer comparisons now produce `CallArgument` constraints, rather than being `Boring`.
- type annotations from turbofished generic arguments are now `TypeAnnotation`s rather than `Boring`. this is tested for functions, but I don't think there's a test for blaming the generic arguments of an ADT constructor so that might need one. ideally [nll/issue-98589-closures-relate-named-regions.rs](https://github.com/rust-lang/rust/blob/master/tests/ui/nll/issue-98589-closures-relate-named-regions.rs) should blame the annotations, since they do get a `TypeAnnotation` category, but all the propagated outlives requirements from the closures have boring categories, due to how propagation is implemented.
- I've added a label to catch when `CopyBound` and `SizedBound` constraints are present but not blamed. these only show up in a few tests, but I think the information is helpful. there's no test for the label pointing out a `SizedBound` constraint, so it might need a test, but there are a couple for `CopyBound`, and the logic is the same.
- I've implemented an alternative way of filtering out uninteresting assignments. it now only catches assignments from argument patterns, since those specifically I think won't be interesting and shouldn't be called "assignments". it's still a bit hacky, but I like it being very clear and specific about what it's making boring.

as far as ordering constraint interest goes, here's some considerations that didn't fit in the comments:
- type annotations can't be more interesting than assignments, or else [those](https://github.com/rust-lang/rust/blob/master/tests/ui/mir/field-projection-mutating-context.stderr) [tests](https://github.com/rust-lang/rust/blob/master/tests/ui/nll/relate_tys/fn-subtype.stderr) for higher-rank subtyping errors will blame the annotation rather than the assignment.
- assigning from a function's returned value shouldn't be more interesting than "usages"; this is what misdirected blame in #132749.
- yields shouldn't be more interesting than type annotations or we'd end up with unhelpful blames like this (from [coroutine/check-resume-ty-lifetimes.rs](https://github.com/rust-lang/rust/blob/master/tests/ui/coroutine/check-resume-ty-lifetimes.rs)):
``` diff
7        LL |             let x: &'static str = yield ();
-           |                    ^^^^^^^^^^^^ type annotation requires that `'1` must outlive `'static`
+           |                                   ^^^^^^^^ yielding this value requires that `'1` must outlive `'static`
```
though the real problem may be that yields need specially worded messages like returns, at which point we could prioritize blaming them. "yielding this value" isn't quite the full story as to what's going on there.

several tests featuring invariance have gone back to blaming returns, and thus have gone back to running afoul of #108636. apart from the message currently being wrong, blaming the return does look better.

the mir-opt tests should be properly blessed now, also!

bors commented:
:umbrella: The latest upstream changes (presumably #134465) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #135151) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dianne commented:
the conflict with #135147 was just the addition of an import near one I changed; no relevant tests were changed.

there was also a conflict with #134627 though; the change to its test output is here: [tests/ui/borrowck/implementation-not-general-enough-ice-133252.stderr](https://github.com/rust-lang/rust/pull/133858/files#diff-d8a6f846b62ef2c026eef21b76c5e12b98445912decfe2d93dbb14ea0ff88ffd). it might be a minor improvement: as the note points out, the `Send` bound on `force_send` is what introduces the `'static` requirement, so having the label point to it makes sense to me.

lcnr commented on tests/ui/borrowck/two-phase-surprise-no-conflict.stderr:
```diff
@@ -115,14 +114,13 @@ error[E0499]: cannot borrow `*reg` as mutable more than once at a time
    |                     -- lifetime `'a` defined here
 ...
+LL |     let reg = mk_reg();
+   |               -------- assignment requires that `reg.sess_mut` is borrowed for `'a`
```
these errors are slightly worse :thinking:

dianne replied:
yeah.. it should probably be pointing to `CapturePass::new` or the coercion (though without the spurious object lifetime default note). unfortunately, I think knowing what to blame there requires keeping better track of variance for diagnostics than we currently do.. I'm not sure the current approach can handle this (though I would love to be wrong)

lcnr commented on tests/ui/lifetimes/copy_modulo_regions.stderr:
```diff
@@ -4,7 +4,10 @@ error: lifetime may not live long enough
-   |     ^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'a` must outlive `'static`
+   |     ^^^^^^^^^^^^^^^^^^^^^
+   |     |
+   |     returning this value requires that `'a` must outlive `'static`
```
the return feels unrelated here :thinking:

dianne replied:
I think this is a consequence of prioritizing blaming returns: [here's what it looks like without that](https://github.com/rust-lang/rust/pull/133858/commits/ac922245f0c86657a56d3ca5eff4a1d1b1fcb0a1#diff-96e1f8b29789b3c4ce2f77a5e0fba248829b97ef9d1ce39e7d2b4aa57b2cf4f0). I'd prefer the return not to be mentioned, but since that'd require more work to get other errors to a good place, I'm happy enough as long as it has a "requirement occurs due to copying this value" label.

lcnr commented on tests/ui/nll/closure-requirements/propagate-approximated-val.stderr:
```diff
@@ -23,17 +23,21 @@ LL | fn test<'a, 'b>(cell_a: Cell<&'a u32>, cell_b: Cell<&'b u32>) {
+   |           --  -- lifetime `'b` defined here
+   |           |
+   |           lifetime `'a` defined here
+LL | /     establish_relationships(cell_a, cell_b, |outlives1, outlives2, x, y| {
+...  |
+LL | |     });
+   | |______^ argument requires that `'a` must outlive `'b`
```
why does this talk about argument while using the span of the function

dianne replied:
that's a quirk of `CallArgument` constraints. they get their span from the `Location` of the call terminator; since they don't store the span of the particular argument, it gets lost in diagnostics. that said, it does look like we have the spans of each argument in `TypeChecker::check_call_inputs`, so they could store that.

lcnr replied:
opened #135238 for that :+1:

lcnr commented on tests/ui/nll/user-annotations/adt-nullary-enums.stderr:
```diff
@@ -1,52 +1,49 @@
+   | |                                         ^^ borrowed value does not live long enough
+LL | |         SomeEnum::SomeOtherVariant::<Cell<&'static u32>>,
+LL | |     );
+   | |_____- argument requires that `c` is borrowed for `'static`
```
same here, surprising span when talking about argument

lcnr commented:
@bors delegate+

feel free to merge with or without looking into my nits

dianne commented:
I've looked into those. I'll leave individual replies for posterity/reference, but in short I think they'd either require bigger or unrelated changes, so I'm inclined to leave them for future work. thanks again!

@bors r=@lcnr

bors commented:
:pushpin: Commit fe8b12f8cfd27984a6102c441b49023d5bce0b02 has been approved by `lcnr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=lcnr fe8b12f8cfd27984a6102c441b49023d5bce0b02 -->
<!-- homu: {"type":"Approved","sha":"fe8b12f8cfd27984a6102c441b49023d5bce0b02","approver":"lcnr","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit fe8b12f8cfd27984a6102c441b49023d5bce0b02 with merge 6afee111c2faf86ba884ea748967130abad37b52...
<!-- homu: {"type":"BuildStarted","head_sha":"fe8b12f8cfd27984a6102c441b49023d5bce0b02","merge_sha":"6afee111c2faf86ba884ea748967130abad37b52"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/12670721853/job/35320128042)
Approved by: lcnr
Pushing 6afee111c2faf86ba884ea748967130abad37b52 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lcnr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/12670721853/job/35320128042"},"merge_sha":"6afee111c2faf86ba884ea748967130abad37b52"} -->

rust-timer commented:
Finished benchmarking commit (6afee111c2faf86ba884ea748967130abad37b52): [comparison URL](https://perf.rust-lang.org/compare.html?start=9c87288a7d2f03625a813df6d3bfe43c09ad4f5a&end=6afee111c2faf86ba884ea748967130abad37b52&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9c87288a7d2f03625a813df6d3bfe43c09ad4f5a&end=6afee111c2faf86ba884ea748967130abad37b52&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.2%  | [0.1%, 21.0%]  | 17    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | 3.2%  | [0.1%, 21.0%]  | 17    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9c87288a7d2f03625a813df6d3bfe43c09ad4f5a&end=6afee111c2faf86ba884ea748967130abad37b52&stat=max-rss)
<details>
<summary>Results (primary 2.7%, secondary -1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.3%  | [2.5%, 6.1%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -1.8% | [-2.1%, -1.4%] | 2     |
| All ❌✅ (primary)                 | 2.7%  | [-2.1%, 6.1%]  | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9c87288a7d2f03625a813df6d3bfe43c09ad4f5a&end=6afee111c2faf86ba884ea748967130abad37b52&stat=cycles:u)
<details>
<summary>Results (primary 4.7%, secondary -1.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.7%  | [0.8%, 20.1%]  | 14    |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [2.1%, 3.9%]   | 5     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.2% | [-4.7%, -1.4%] | 17    |
| All ❌✅ (primary)                 | 4.7%  | [0.8%, 20.1%]  | 14    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9c87288a7d2f03625a813df6d3bfe43c09ad4f5a&end=6afee111c2faf86ba884ea748967130abad37b52&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 764.901s -> 763.722s (-0.15%)
**Artifact size**: 325.67 MiB -> 325.68 MiB (0.01%)

Kobzol commented:
Oh, that perf. regression looks quite bad! Probably it was introduced after some later changes, because a previous perf. run did not show any regressions. Looks like some MIR queries were executed [more often](https://perf.rust-lang.org/detailed-query.html?commit=6afee111c2faf86ba884ea748967130abad37b52&benchmark=serde_derive-1.0.136-check&scenario=incr-patched%3A+println&base_commit=9c87288a7d2f03625a813df6d3bfe43c09ad4f5a&sort_idx=-11).

If the fix is obvious, it would be great to send a PR with a fix. If not, I'd suggest reverting this first, since the regression seems to affect real world crates.

lqd commented:
I'll prepare a revert that we can land in case the fix is not obvious or @dianne doesn't have time to take a look.

dianne commented:
I have an idea of what it could be.. kind of. I'm not confident, and I can't seem to get the profiler running locally to be sure, but I've opened a PR: #135273.

Since the regressions come up during incremental compilation and seem to be query-system-related, my guess is that I committed a query sin, resulting in lots, lots more executions of `mir_borrowck`. I made a tweak to MIR typeck to treat certain MIR assignments that weren't written by the user as not written by the user. In order to avoid adding more information to the MIR or tracking more things during pattern lowering, I wrote a bit of a hack that queried the HIR. If that required computing something not in the incremental cache, or if it somehow clobbered/invalidated it, that could explain the huge number of additional queries. I'm not too familiar with the query system, so I could be way off base, but my other guesses are all scarier and less trivial, so I'm hoping it's this.

lqd commented:
The perf regression was fixed in https://github.com/rust-lang/rust/pull/135273#issuecomment-2582715034
marking as triaged @rustbot label: +perf-regression-triaged
